runtime/compiler/optimizer/JProfilingRecompLoopTest.cpp

```diff
@@ -0,0 +1,267 @@
+         return true;   
+      }
+   return false;
+   }
+
+int32_t
+TR_JProfilingRecompLoopTest::perform()
+   {
+   if (comp()->getOption(TR_EnableJProfiling))
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has been enabled, run JProfiling\n");
+      }
+   else if (comp()->getProfilingMode() == JProfiling)
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has been enabled for profiling compilations, run JProfilingRecompLoopTest\n");
+      }
+   else
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has not been enabled, skip JProfilingRecompLoopTest\n");
+      return 0;
+      }
+   RecompilationTestLocations testLocations(RecompilationTestLocationAllocator(comp()->trMemory()->currentStackRegion()));
+   TR::TreeTop *cursor = comp()->getStartTree();
+   TR::CFG *cfg = comp()->getFlowGraph();
+   TR::Block *currentBlock = NULL;
+   TR::list<TR_ByteCodeInfo, TR::Region&> addedLocationBCIList(comp()->trMemory()->currentStackRegion());
+   while (cursor)
+      {
+      if (cursor->getNode()->getOpCodeValue() == TR::BBStart)
+         {
+         currentBlock = cursor->getNode()->getBlock();
+         /**
+           * As we are already walking down the tree tops, we can get the enclosing block by tracking TR::BBStart nodes and currentBlock contains
+           * this information. 
+           * We also keep local list of ByteCodeInfo for each test locations (asyncchecknode) in the extended basic blocks.
+           * This can avoid adding multiple tests for multiple locations with same byte code info in extended basic block.
+           * As soon as we encounter a block which is not extenstion of previous block, we clear the list of byte code info.
+           */    
+         if (!currentBlock->isExtensionOfPreviousBlock() && !addedLocationBCIList.empty())
+            addedLocationBCIList.clear();
+         }
+      else if (cursor->getNode()->getOpCodeValue() == TR::asynccheck)
+         {
+         TR_ASSERT_FATAL(currentBlock != NULL,"We should have encountered BBStart before and should have the enclosing block");
+         if (currentBlock->getStructureOf()->getContainingLoop() != NULL)
+            {
+            TR_ByteCodeInfo bci = cursor->getNode()->getByteCodeInfo();
+            // If the list of TR_ByteCodeInfo used to track the tree tops in extended basic blocks is empty or the bytecode info
+            // of current asynccheck node does not match with any tracked bytecode info, we will add the test otherwise skip it.   
+            if (addedLocationBCIList.empty() || !isByteCodeInfoInCurrentTestLocationList(bci, addedLocationBCIList))
+               {
+               addedLocationBCIList.push_back(bci);
+               int32_t loopDepth = getLoopNestingDepth(comp(), currentBlock);
+               testLocations.push_back(std::make_pair(std::make_pair(cursor, currentBlock),loopDepth));
```

Comment:
why do we have to store this and can't we just get it from structure when needed?

Reply:
Once we invalidate the structure, we lost the information about loop nesting. That is why we have to store the loopDepth before going on and splitting blocks.